docs: add Understanding Kubernetes Networking pillar page (#961)
cilium/cilium.io: src/posts/2026-04-25-understanding-kubernetes-networking/index.md

The goal is for each pod to have an IP in a flat shared networking namespace that has full communication with other pods across the network. IP-per-pod creates a clean, backward-compatible model where pods can be treated much like VMs or physical hosts from the perspectives of port allocation, networking, naming, service discovery, load balancing, application configuration, and migration.

- No Internal NAT: Internal pods communication doesn’t use Network Address translation.
- The same network namespace for containers inside the same pod: containers in the same pod share an IP address and share port space.
- Services as stable access points: services provide virtual IPs that map to backend pods.
- Overlay or underlay networks: Some CNIs use overlay networks to encapsulate traffic. Others rely on Layer 3 routing to encapsulate traffic.

Read More: [https://isovalent.com/blog/post/what-is-kubernetes-networking/](https://isovalent.com/blog/post/what-is-kubernetes-networking/)

### Core Networking Requirements

- Highly coupled Container to Container communications: containers within the same pod share the same network namespace. Communication is done via localhost.

> - Pod-to-Pod communications: _How does Kubernetes Pod-to-Pod communication work?_

Reviewer: ```suggestion
- Pod-to-Pod communications:
```

  Every pod can communicate directly with any other pod across nodes without NAT since each pod is assigned a real IP. This enables all naming or discovery mechanisms to work out of the box.

> - Pod to Service communications: A service is an abstraction of a group of pods. Services are assigned Virtual IPs proxied to pods in a service, since pods are ephemeral workloads, don’t need to care about pod IP changes; Kubernetes automatically tracks these via Service EndpointSlice. Kube Proxy keeps track of the iptables to track VIP services; Cilium does this smartly by replacing iptables with eBPF. With eBPF, packet decisions happen at the earliest hook, and faster network policy evaluation with a low CPU overhead.

Reviewer: ```suggestion
- Pod to Service communications: A service is an abstraction of a group of pods. Services are assigned Virtual IPs proxied to pods in a service. Since pods are ephemeral workloads, they don’t need to care about pod IP changes; Kubernetes automatically tracks these via Service EndpointSlice. Kube-proxy keeps track of the iptables to track VIP services; Cilium does this smartly by replacing iptables with eBPF. With eBPF, packet decisions happen at the earliest hook, and faster network policy evaluation with a low CPU overhead.
```

- External to Internal communications: Exposing and accessing services, such as external services or databases, from outside the Kubernetes cluster.

Read More: [https://github.com/kubernetes/design-proposals-archive/blob/main/network/networking.md](https://github.com/kubernetes/design-proposals-archive/blob/main/network/networking.md).

### Why does traditional networking fail in a dynamic, containerized world?

> Containerised environments run resources with an ephemeral nature, and hence it’s hard to have the same traditional networking principles.

Reviewer: ```suggestion
Containerised environments often run resources with an ephemeral nature, and traditional networking technologies struggle to keep up with quickly changing environments.
```

eBPF allows Cilium to embed logic or programs into the Linux kernel dynamically.

The following are key capabilities that Cilium offers to support the Kubernetes networking model.

1. Performance and Scalability: Cilium avoids old routing technologies like iptables or IPVS that limit performance and scalability.
2. Identity-based security: Instead of managing security via ephemeral IP addresses, Cilium assigns a unique security identity to pods based on metadata. Most CNIs can enforce network policies at L3 and L4, but Cilium extends this capability to provide L7 control capabilities too.
3. Multi-cluster connectivity: Cilium provides the ability to connect multiple clusters across different clouds.
4. Transparent Encryption: Kubernetes lacks native pod-to-pod encryption. Two common solutions to this problem are embedding encryption within the application or using a service mesh. Embedding encryption within the app is complex and requires application and security expertise. Cilium provides a straightforward solution for enabling the encryption of all node-to-node traffic with just one switch, no application changes, or additional proxies.

> 5. Observability: Cilium's Metrics and Tracing export feature provides a seamless and integrated solution, empowering users to monitor, analyze, and optimize their Kubernetes environments with ease. By leveraging the power of Prometheus metrics, combined with Hubble's network behavior insights, Cilium enables users to gain deep visibility into their applications and network while simplifying the setup and configuration process.

Reviewer: ```suggestion
5. Observability: Cilium's Hubble empowers users to monitor, analyze, and optimize their Kubernetes networking environments with ease. With both a CLI and service map, Hubble helps quickly identify problems in the network.
```

> ![Observability with Cilium](./images/image1.png)

Reviewer: I would switch this to an image of Hubble, maybe the service map

Author: Noted, let me review again the images

Author: switched to a new Hubble image

- The Outer Header: The original packet is wrapped (encapsulated) in a UDP header that uses the Node’s IP address as the source and destination.

Because the physical network only sees Node-to-Node traffic, connectivity is highly resilient.

#### The Trade-offs

While Overlay mode introduces a small amount of encapsulation overhead (extra bytes in the packet header), it provides operational simplicity. In this mode, Cilium creates a virtual network that sits on top of the existing physical infrastructure.

Read More: [https://docs.cilium.io/en/latest/network/concepts/routing/\#encapsulation](https://docs.cilium.io/en/latest/network/concepts/routing/#encapsulation)

| Feature     | Native Routing                                    | Encapsulated (Overlay)            |
| :---------- | :------------------------------------------------ | :-------------------------------- |
| Complexity  | High (requires configuration on physical network) | Lower (Works out of the box)      |
| Performance | High (no overhead)                                | Moderate (encapsulation overhead) |
| Visibility  | Pod IPs visible to the network                    | Pod IPs hidden from the network   |

> ![Encapsulation vs Native Routing](./images/image3.png)

Reviewer: Where did you get this image from, it doesn't explain the difference

- DNS proxy: Cilium uses a DNS proxy to listen for DNS traffic. This adds a layer of granularity for security enforcement. This allows you to write security policies based on FQDN and not only IPs.

# V. Summary

The Kubernetes networking model requires an understanding of the shift from static, physical infrastructure to dynamic, software-defined environments. While the fundamental rules, such as the IP-per-Pod model and Flat Network Structure, provide the necessary consistency for distributed applications, traditional networking tools often struggle to keep pace with the ephemeral nature of containers.

Cilium doesn't just fulfill the CNI specification; it has become the industry-standard foundation for cloud native infrastructure by moving networking logic directly into the Linux kernel via eBPF. This shift has led to massive adoption across the ecosystem, with major cloud providers (including Google, AWS, and Azure) choosing Cilium as their default networking layer. As a CNCF-graduated project with a vibrant, global community, Cilium has effectively set the new benchmark for networking, security, observability, and scale in the modern enterprise.

> As organizations increasingly deploy compute-intensive AI and machine learning workloads across hybrid environments, the requirements for security, observability, and operational scale have become non-negotiable. Cilium has set a standard in Kubernetes networking by leveraging eBPF to provide the high-performance, identity-based networking infrastructure necessary to meet these modern demands.

Reviewer: ```suggestion
```
